File: typescript/src/examples.ts
/**
 * Typed example calls to the anera.markets public HTTP API (Node 18+ fetch).
 *
 * Set ANERA_MARKETS_API_BASE_URL to your API origin (scheme + host, no trailing slash).
 */

import type {
  CompanyItem,
  ModelItem,
  ResourceType,
  RevenueQueryParams,
  RevenueResponse,
  TokenFactoryItem,
  TokenUtilisationQueryParams,
  TokenUtilisationResponse,
} from "./types.js";

export type {
  CompanyItem,
  ModelItem,
  ResourceType,
  RevenueQueryParams,
  RevenueResponse,
  RevenueRow,
  TokenFactoryItem,
  TokenType,
  TokenUtilisationQueryParams,
  TokenUtilisationResponse,
  TokenUtilisationRow,
} from "./types.js";

function baseUrl(): string {
  const base = (process.env.ANERA_MARKETS_API_BASE_URL ?? "").trim().replace(/\/$/, "");
  if (!base) {
    console.error(
      "Set ANERA_MARKETS_API_BASE_URL to the API origin, e.g. ANERA_MARKETS_API_BASE_URL=https://api.example.com",
    );
    process.exit(1);
  }
  return base;
}

function buildQuery(params: Record<string, string | undefined>): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null) search.set(k, String(v));
  }
  const q = search.toString();
  return q ? `?${q}` : "";
}

async function getJson<T>(path: string, params: Record<string, string | undefined> = {}): Promise<T> {
  const url = `${baseUrl()}${path}${buildQuery(params)}`;
  const res = await fetch(url, { headers: { Accept: "application/json" } });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`HTTP ${res.status}: ${text}`);
  }
  return res.json() as Promise<T>;
}

export function getModels(): Promise<ModelItem[]> {
  return getJson<ModelItem[]>("/api/v1/public/models");
}

export function getTokenFactories(): Promise<TokenFactoryItem[]> {
  return getJson<TokenFactoryItem[]>("/api/v1/public/token-factories");
}

export function getCompanies(): Promise<CompanyItem[]> {
  return getJson<CompanyItem[]>("/api/v1/public/companies");
}

export function getRevenue(
  resourceType: ResourceType,
  params: RevenueQueryParams = {},
): Promise<RevenueResponse> {
  const { timestamp, resource_id: resourceId } = params;
  return getJson<RevenueResponse>(`/api/v1/public/revenue/${resourceType}`, {
    ...(timestamp !== undefined ? { timestamp } : {}),
    ...(resourceId !== undefined ? { resource_id: resourceId } : {}),
  });
}

export function getTokenUtilisation(
  resourceType: ResourceType,
  params: TokenUtilisationQueryParams = {},
): Promise<TokenUtilisationResponse> {
  const {
    timestamp,
    resource_id: resourceId,
    token_type: tokenType = "total",
  } = params;
  return getJson<TokenUtilisationResponse>(`/api/v1/public/token-utilisation/${resourceType}`, {
    token_type: tokenType,
    ...(timestamp !== undefined ? { timestamp } : {}),
    ...(resourceId !== undefined ? { resource_id: resourceId } : {}),
  });
}

async function main(): Promise<void> {
  console.log("Models (first 3):");
  const models = await getModels();
  console.log(JSON.stringify(models.slice(0, 3), null, 2));

  console.log("\nToken factories (first 3):");
  const factories = await getTokenFactories();
  console.log(JSON.stringify(factories.slice(0, 3), null, 2));

  console.log("\nCompanies (first 3):");
  const companies = await getCompanies();
  console.log(JSON.stringify(companies.slice(0, 3), null, 2));

  console.log("\nRevenue by model (latest day, sample):");
  const revenue = await getRevenue("model");
  console.log(
    JSON.stringify(
      {
        resource_type: revenue.resource_type,
        timestamp: revenue.timestamp,
      },
      null,
      2,
    ),
  );
  console.log(JSON.stringify((revenue.items ?? []).slice(0, 5), null, 2));

  console.log("\nToken utilisation by company (total, latest day, sample):");
  const util = await getTokenUtilisation("company", { token_type: "total" });
  console.log(
    JSON.stringify(
      {
        resource_type: util.resource_type,
        timestamp: util.timestamp,
        token_type: util.token_type,
      },
      null,
      2,
    ),
  );
  console.log(JSON.stringify((util.items ?? []).slice(0, 5), null, 2));
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});

File: python/general-examples/examples.py
"""
Example calls to the anera.markets public HTTP API.

Set ANERA_MARKETS_API_BASE_URL to your API origin (scheme + host, no trailing slash).
Example: export ANERA_MARKETS_API_BASE_URL=https://api.example.com
"""

from __future__ import annotations

import json
import os
from typing import Any, Literal

import requests

ResourceType = Literal["token-factory", "model", "company"]
TokenType = Literal["total", "prompt", "completion", "reasoning"]


def _base_url() -> str:
    base = os.environ.get("ANERA_MARKETS_API_BASE_URL", "").strip().rstrip("/")
    if not base:
        raise SystemExit(
            "Set ANERA_MARKETS_API_BASE_URL to the API origin, e.g. "
            "export ANERA_MARKETS_API_BASE_URL=https://api.example.com"
        )
    return base


def _get(path: str, params: dict[str, Any] | None = None) -> Any:
    url = f"{_base_url()}{path}"
    r = requests.get(url, params=params or {}, timeout=60)
    r.raise_for_status()
    return r.json()


def get_models() -> list[dict[str, Any]]:
    return _get("/api/v1/public/models")


def get_token_factories() -> list[dict[str, Any]]:
    return _get("/api/v1/public/token-factories")


def get_companies() -> list[dict[str, Any]]:
    return _get("/api/v1/public/companies")


def get_revenue(
    resource_type: ResourceType,
    *,
    timestamp: str | None = None,
    resource_id: str | None = None,
) -> dict[str, Any]:
    params: dict[str, Any] = {}
    if timestamp is not None:
        params["timestamp"] = timestamp
    if resource_id is not None:
        params["resource_id"] = resource_id
    return _get(f"/api/v1/public/revenue/{resource_type}", params=params)


def get_token_utilisation(
    resource_type: ResourceType,
    *,
    timestamp: str | None = None,
    resource_id: str | None = None,
    token_type: TokenType = "total",
) -> dict[str, Any]:
    params: dict[str, Any] = {"token_type": token_type}
    if timestamp is not None:
        params["timestamp"] = timestamp
    if resource_id is not None:
        params["resource_id"] = resource_id
    return _get(f"/api/v1/public/token-utilisation/{resource_type}", params=params)


def main() -> None:
    print("Models (first 3):")
    models = get_models()
    print(json.dumps(models[:3], indent=2))

    print("\nToken factories (first 3):")
    factories = get_token_factories()
    print(json.dumps(factories[:3], indent=2))

    print("\nCompanies (first 3):")
    companies = get_companies()
    print(json.dumps(companies[:3], indent=2))

    print("\nRevenue by model (latest day, sample):")
    revenue = get_revenue("model")
    print(json.dumps({k: revenue[k] for k in ("resource_type", "timestamp")}, indent=2))
    items = revenue.get("items") or []
    print(json.dumps(items[:5], indent=2))

    print("\nToken utilisation by company (total, latest day, sample):")
    util = get_token_utilisation("company", token_type="total")
    print(json.dumps({k: util[k] for k in ("resource_type", "timestamp", "token_type")}, indent=2))
    uitems = util.get("items") or []
    print(json.dumps(uitems[:5], indent=2))


if __name__ == "__main__":
    main()
